projects/permissions/src/services/permissions.service.ts

```diff
@@ -1,6 +1,51 @@
-import {Injectable} from '@angular/core';
+import {Inject, Injectable} from '@angular/core';
+import {from, fromEvent, Observable} from 'rxjs';
+import {shareReplay, switchMap, tap} from 'rxjs/operators';
+import {PERMISSIONS} from '../tokens/permissions';
+import {PERMISSIONS_SUPPORT} from '../tokens/permissions-support';
+
+export type PermissionsQueryArgs = Parameters<typeof Permissions.prototype.query>[0];
 
 @Injectable({
     providedIn: 'root',
 })
-export class PermissionsService {}
+export class PermissionsService {
+    constructor(
+        @Inject(PERMISSIONS) private readonly permissions: Permissions,
+        @Inject(PERMISSIONS_SUPPORT) private permissionsSupported: boolean,
+    ) {}
+
+    state(name: PermissionName): Observable<PermissionState>;
+    state(descriptor: PermissionsQueryArgs): Observable<PermissionState>;
+    state(
+        nameOrDescriptor: PermissionName | PermissionsQueryArgs,
+    ): Observable<PermissionState> {
+        const descriptor: PermissionDescriptor =
+            typeof nameOrDescriptor === 'string'
+                ? {name: nameOrDescriptor}
+                : nameOrDescriptor;
+
+        return new Observable<PermissionState>(subscriber => {
+            if (!this.permissionsSupported) {
+                subscriber.error('Permissions is not supported in your browser');
+
+                return;
+            }
+
+            const sub = from(this.permissions.query(descriptor))
+                .pipe(
+                    tap(permissionStatus => subscriber.next(permissionStatus.state)),
```

Comment:
IMHO, that will be more readable (the tap operator removed)
```
switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(startWith(permissionStatus.state))),
```

Reply:
I don't know. We're getting into multiple pipe nesting. Besides, in order to make that work, the code should be something like this:
```ts
const sub = from(this.permissions.query(descriptor))
    .pipe(
      switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(
        map(ev => (ev.target as PermissionStatus).state),
        startWith(permissionStatus.state),
      )),
    )
    .subscribe(
        state => subscriber.next(state),
        error => subscriber.error(error),
    );
```
I'd argue this is less readable.

Reply:
Sorry, I meant
```typescript
switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(startWith(permissionStatus))),
```

Reply:
that's basically what I have in my example there. Except that I can't use the `startWith` as a first operator in the inner pipe because `fromEvent(permissionStatus, 'change')` returns an object of type `Event`. So I can't "start it with" an object of type `PermissionState`. There needs to be a `map` in between to match types.

Unless I misunderstand something. Please provide a full snippet.

Reply:
oh, I see the difference. Still though there needs to be a `map`:
```ts
const sub = from(this.permissions.query(descriptor))
  .pipe(
    switchMap(permissionStatus => fromEvent(permissionStatus, 'change').pipe(
      map(ev => (ev.target as PermissionStatus)),
      startWith(permissionStatus),
    )),
  )
  .subscribe(
      permissionStatus => subscriber.next(permissionStatus.state),
      error => subscriber.error(error),
  );
```

Reply:
You're right. I read inattentively